### PR TITLE
parse-type: don't inline type application

### LINKED ISSUE #1234: Typed Racket hangs on a simple program
### What version of Racket are you using?
v8.4 [cs] and v8.5 [cs]
### What program did you run?
```Racket
#lang typed/racket/base

(provide (all-defined-out))

(define-type (Formula a)
  (U
   True
   False
   (Atom a)
   (Not a)
   (And a)
   (Or a)
   (Imp a)
   (Iff a)
   (Forall a)
   (Exists a)))

(struct False () #:transparent)
(struct True () #:transparent)
(struct (a) Atom ([f : a]) #:transparent)
(struct (a) Not ([f : (Formula a)]) #:transparent)
(struct (a) And ([l : (Formula a)] [r : (Formula a)]) #:transparent)
(struct (a) Or ([l : (Formula a)] [r : (Formula a)]) #:transparent)
(struct (a) Imp ([pre : (Formula a)] [post : (Formula a)]) #:transparent)
(struct (a) Iff ([l : (Formula a)] [r : (Formula a)]) #:transparent)
(struct (a) Forall ([v : String] [f : (Formula a)]) #:transparent)
(struct (a) Exists ([v : String] [f : (Formula a)]) #:transparent)
```
### What should have happened?
I'd expect this program to type check quickly.

### PR DESCRIPTION
**Pull request by capfredf**
This pull request includes two major changes:
1. type applications are not resolved until the results are used (also closes #1234)
2. streamline the handling of variances

### CODE REVIEW

File: typed-racket-lib/typed-racket/rep/type-constr.rkt [hunk 1]
```diff
@@ -49,7 +49,9 @@
       (when (or (< len arity)
                 (and (> len arity) (not kind*?)))
         (raise (make-exn:fail:contract:arity:type-constructor
-                (format "number of expected type arguments: ~a~n given: ~a" arity len)
+                (format "~a: number of expected type arguments: ~a~n given: ~a"
+                        (object-name real-trep-constr)
+                        arity (cons len args))
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/rep/type-constr.rkt:54]: add a field for args in the error report

File: typed-racket-lib/typed-racket/env/type-alias-helper.rkt [hunk 2]
```diff
@@ -204,13 +204,19 @@
...
        ;; id can be a simple abbreviation for another type constructor
        (define rv (parse-type-or-type-constructor type-stx))
-       ((if (TypeConstructor? rv)
-            register-type-constructor!
-            register-resolved-type-alias) id rv)])
+       (match rv
+         [(struct* TypeConstructor ([arity arity]
+                                    [kind*? kind*?]))
+          (register-type-constructor! id rv)
+          (add-constant-variance! id (build-list (if (zero? arity) 1
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/env/type-alias-helper.rkt:216]: comment?

File: typed-racket-lib/typed-racket/env/init-envs.rkt [hunk 3]
```diff
@@ -364,9 +364,10 @@
     ;; Most Top types are in the predefined table, the ones here
     ;; are not
     [(StructTop: name) `(make-StructTop ,(type->sexp name))]
-    [(TypeConstructor constr arity kind*? productive?)
+    [(TypeConstructor constr arity kind*? productive? variance)
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/env/init-envs.rkt:367]: a type can have multiple variances.

File: typed-racket-lib/typed-racket/env/init-envs.rkt [hunk 4]
```diff
@@ -489,10 +490,11 @@
    kind-env-map
    (lambda (id v)
      ;; TODO: turn this into a function
-     (match-define (TypeConstructor constr arity kind*? productive?) v)
+     (match-define (TypeConstructor constr arity kind*? productive? variance) v)
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/env/init-envs.rkt:493]: same here

File: typed-racket-lib/typed-racket/private/parse-type.rkt [hunk 5]
```diff
@@ -605,7 +607,7 @@
      (-val (syntax->datum #'t))]))
 
 (struct parsing-mode () #:transparent)
-(struct synth-mode parsing-mode (name args same-component-pred) #:transparent)
+(struct synth-mode parsing-mode (name args same-component-pred constructor-variances) #:transparent)
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/private/parse-type.rkt:610]: constructor-variances is not used.

**Comment by jackfirth** [comment]
@capfredf No, it should ignore them but it wasn't doing that. I pushed a fix to Resyntax to only analyze added and modified files.

**Review by samth — APPROVED** [review]

**Pull request merged**